typings/index.d.ts

```diff
@@ -14,11 +14,85 @@ declare module 'discord.js' {
   import { ChildProcess } from 'child_process';
   import { EventEmitter } from 'events';
   import { PathLike } from 'fs';
+  import { Response } from 'node-fetch';
   import { Readable, Stream, Writable } from 'stream';
+  import { URLSearchParams } from 'url';
   import * as WebSocket from 'ws';
 
   export const version: string;
 
+  //#region Rest
+
+  class AsyncQueue {
+    // tslint:disable-next-line:ban-types
+    private readonly promises: { promise: Promise<void>; resolve(): void }[];
+    public remaining: number;
+    public wait(): Promise<void>;
+    private shift(): void;
+  }
+
+  class RequestHandler {
+    private readonly manager: RESTManager;
+    private readonly queue: AsyncQueue;
+    public reset: number;
+    public remaining: number;
+    public limit: number;
+    public retryAfter: number;
+    public limited: boolean;
+    private readonly inactive: boolean;
+
+    private execute(): Record<string, unknown> | Buffer;
+    public push(): Record<string, unknown> | Buffer;
+  }
+
+  interface RequestOptions {
+    query?: URLSearchParams | Record<string, string | string[]>;
+    versioned?: boolean;
+    auth?: boolean;
+    reason?: string;
+    headers?: {};
+    data?: {};
+  }
+
+  interface RouteBuilderMethods {
+    get(opts?: RequestOptions): Promise<unknown>;
+    post(opts?: RequestOptions): Promise<unknown>;
+    delete(opts?: RequestOptions): Promise<unknown>;
+    patch(opts?: RequestOptions): Promise<unknown>;
+    put(opts?: RequestOptions): Promise<unknown>;
+  }
+
+  type RouteBuilder = Record<string, RouteBuilderMethods & ((...args: string[]) => RouteBuilderMethods)>;
```

Comment:
```suggestion
type HttpMethod = 'get' | 'post' | 'delete' | 'patch' | 'put';

export type RouteBuilder = Record<
    HttpMethod,
    <T>(options?: RequestOptions) => Promise<T>
> &
    {
        [k in string]: RouteBuilder;
    } &
    ((...args: string[]) => RouteBuilder);
```
With your implementation you cannot do more that 1 call (eg `api.foo.goo.get()`).
Also I added a generic type for http methods but I don't know if it's a good idea or not.

[playground](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgEoQI4FcIGcwDyADmMAPYi7IDeyAvgLABQzYAnkSgBJhhECyEMAAsyAE2QBeZAHIA5kJnIAPrKJl8S1TLEQANkIha1cMAmHGZRLGBkBuZswgAPdVDDJ2nNGRsQAQljAerpQUmgQCGRQYgA8zMiJyDx8giLiADQJSbEAKgB8ABRkJOSUAPwAXBHYeISlFLgAlFL5yAAKUGQAtsC4EHn5zG0AZNmJ1ONJiQDaANbIoMj4UKByALrVqL6QgcGhDkzTdMhjR0mFhQB0N3BQcrjVK2sz6y2Sbdt+eyHQTYfMAweLp+LY7AJBX5haTUE5wKhwEBsAFMEGQK4wMhkK5yLE4oSFf6OJgAejJ5IplKplOJoEgsEQKC+u0hoTSojEVEm50SCjAxRIuCqNRw+GIpEaTWqnR6fQGWBAcxAZAA7iB8odpup8AKwEKtphRfUJZQpR0ur1+rEFUrVerNUldECILr9SK6uKys1pRa5dbFcq1RqpkRTOZXcL0LUxQ1TT7ZVabYH7SGbBGDdHjV6zTLLfKA3bg0xGCwmF4UAB9ZkQ-bQcLoKIxWLPEByDI+b6s6Ds8RUEbIS43K53B5PMCrVuvd6fcE-NlCDnNIuAoTIOBEYDVKuzrvQmhwhFIlHr4AYvG47F8wl2ZBkmWcdxsWQXpRiMh4ZDKjwuPoeCieDgUBkas527Bde1OAdrlue5HmWccXjeVoOxZWsoB7TkmhkIA)

Reply:
Good catch! As for the generic type, see https://github.com/discordjs/discord.js/pull/5150#discussion_r548905773.